### PR TITLE
Crossfaction BG | AzerothCore

### PR DESCRIPTION
**Pull request by vhiperdev**
Implementing CrossfactionBG

Cfbg worked from sunwellcore and fused with azerothcore

Target branch(es):
cfbg

### CODE REVIEW

**Comment by talamortis** [comment]
Please fix travis ```
[1m/home/travis/build/azerothcore/azerothcore-wotlk/src/server/game/Battlegrounds/BattlegroundQueue.cpp:199:24: [0m[0;1;31mfatal error: [0m[1munused variable 'MaxPlayers' [-Wunused-variable][0m
                uint32 MaxPlayers = MinPlayers * 2;
```

**Comment by milestorme** [comment]
tested battleground horde v horde , battleground alliance v alliance, tested arena 1v1 Horde vs Horde , 1v1 Alliance vs Alliance,  2v2 Horde vs Horde , 2v2 Alliance vs Alliance.
AzerothCore rev. 0784e1e6c532+ 2019-06-23 02:07:06 +0100 (pr-1996 branch) (Unix, Release)

no errors, no crash, no build warnings

**Comment by Sdyess** [comment]
Planning to review the code sometime next week.

**Comment by spinzaku** [comment]
Does this have raid check? like 10 hordies make a raid group and queue up bg. Mighy result not being able to attack each other in bg

File: src/server/game/Battlegrounds/Battleground.cpp [hunk 1]
```diff
@@ -264,7 +268,7 @@ void Battleground::Update(uint32 diff)
             else
             {
                 _ProcessResurrect(diff);
-                if (sBattlegroundMgr->GetPrematureFinishTime() && (GetPlayersCountByTeam(TEAM_ALLIANCE) < GetMinPlayersPerTeam() || GetPlayersCountByTeam(TEAM_HORDE) < GetMinPlayersPerTeam()))
```
Comment (Viste) [inline src/server/game/Battlegrounds/Battleground.cpp]:
```
if (sWorld->getBoolConfig(BATTLEGROUND_CROSSFACTION_ENABLED)
{
    if (sBattlegroundMgr->GetPrematureFinishTime() && (GetAlliancePlayersCount() == 0 || GetHordePlayersCount() == 0)))
        _ProcessProgress(diff);
}
else
{
    if (sBattlegroundMgr->GetPrematureFinishTime() && (GetPlayersCountByTeam(TEAM_ALLIANCE) < GetMinPlayersPerTeam() || GetPlayersCountByTeam(TEAM_HORDE) < GetMinPlayersPerTeam()))
        _ProcessProgress(diff);
}
```
or better

```
#ifdef CFBG
if (sBattlegroundMgr->GetPrematureFinishTime() && (GetAlliancePlayersCount() == 0 || GetHordePlayersCount() == 0))
#else
if (sBattlegroundMgr->GetPrematureFinishTime() && (GetPlayersCountByTeam(TEAM_ALLIANCE) < GetMinPlayersPerTeam() || GetPlayersCountByTeam(TEAM_HORDE) < GetMinPlayersPerTeam()))
#endif
```

Comment (Viste) [reply src/server/game/Battlegrounds/Battleground.cpp]: end etc over pr, no need change logic of queue when CFBG not enable,

**Comment by FrancescoBorzi** [comment]
First of all big thank you @vhiperdev and thanks to all the people who are reviewing the code and testing this PR. Your feedback is very precious for us.

I wanted to answer some common questions about this PR: many people are wondering whether this will be merged and/or become a module.

Here's my personal opinion (open to discussion):

- If this works fine it can easily be merged to a `cfbg` branch that can be updated with `master` from time to time. This is the easiest approach.

- If we are 100% that when the CFBG feature is disabled (`Battleground.Crossfaction.Enabled = 1`) there  are no side effects at all with everything else, we can even consider to merge this directly into `master`. This would be a better approach since we don't have to maintain another branch (and avoid things like merge conflicts from time to time etc...)

- If someone wants to convert this feature into a module, **that would be the best approach** for many reasons.

File: src/server/game/Battlegrounds/Battleground.cpp (hunk 1 above)
Comment (vhiperdev) [reply src/server/game/Battlegrounds/Battleground.cpp]: Thanks for your feedback

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -7138,28 +7166,28 @@ void Player::RewardReputation(Unit* victim, float rate)
...
     {
-        int32 donerep1 = CalculateReputationGain(REPUTATION_SOURCE_KILL, victim->getLevel(), Rep->RepValue1, ChampioningFaction ? ChampioningFaction : Rep->RepFaction1);
+        int32 donerep1 = CalculateReputationGain(REPUTATION_SOURCE_KILL, victim->getLevel(), Rep->RepValue1, ChampioningFaction ? ChampioningFaction : repfaction1);
         donerep1 = int32(donerep1 * rate);
 
         FactionEntry const* factionEntry1 = sFactionStore.LookupEntry(ChampioningFaction ? ChampioningFaction : Rep->RepFaction1);
         uint32 current_reputation_rank1 = GetReputationMgr().GetRank(factionEntry1);
-        if (factionEntry1)
-            GetReputationMgr().ModifyReputation(factionEntry1, donerep1, bool(current_reputation_rank1 > Rep->ReputationMaxCap1));
+		if (factionEntry1)
```
Comment (poszer) [inline src/server/game/Entities/Player/Player.cpp]: tabs

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -7138,28 +7166,28 @@ void Player::RewardReputation(Unit* victim, float rate)
...
+			GetReputationMgr().ModifyReputation(factionEntry1, donerep1, bool(current_reputation_rank1 > Rep->ReputationMaxCap1));
```
Comment (poszer) [inline src/server/game/Entities/Player/Player.cpp]: same, tabs

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -7138,28 +7166,28 @@ void Player::RewardReputation(Unit* victim, float rate)
...
-        int32 donerep2 = CalculateReputationGain(REPUTATION_SOURCE_KILL, victim->getLevel(), Rep->RepValue2, ChampioningFaction ? ChampioningFaction : Rep->RepFaction2);
+        int32 donerep2 = CalculateReputationGain(REPUTATION_SOURCE_KILL, victim->getLevel(), Rep->RepValue2, ChampioningFaction ? ChampioningFaction : repfaction2);
         donerep2 = int32(donerep2 * rate);
 
-        FactionEntry const* factionEntry2 = sFactionStore.LookupEntry(ChampioningFaction ? ChampioningFaction : Rep->RepFaction2);
+        FactionEntry const* factionEntry2 = sFactionStore.LookupEntry(ChampioningFaction ? ChampioningFaction : repfaction2);
         uint32 current_reputation_rank2 = GetReputationMgr().GetRank(factionEntry2);
-        if (factionEntry2)
-            GetReputationMgr().ModifyReputation(factionEntry2, donerep2, bool(current_reputation_rank2 > Rep->ReputationMaxCap2));
+		if (factionEntry2)
```
Comment (poszer) [inline src/server/game/Entities/Player/Player.cpp]: tabs

File: src/server/game/Entities/Player/Player.cpp [hunk 5]
```diff
@@ -7138,28 +7166,28 @@ void Player::RewardReputation(Unit* victim, float rate)
...
+			GetReputationMgr().ModifyReputation(factionEntry2, donerep2, bool(current_reputation_rank2 > Rep->ReputationMaxCap2));
```
Comment (poszer) [inline src/server/game/Entities/Player/Player.cpp]: tabs

**Comment by BarbzYHOOL** [comment]
@vhiperdev @Shard-MW bug with the battlegrounds? see micrah's answer

**Comment by Winfidonarleyan** [comment]
And so, go
1. Need rework all code. To do as little as possible of the functions in the core, so all the functions were only in the class CFBG.
2. There is no need to do any actions not on the battleground, why overload the core?

I know what I'm saying, I have my own module

**Comment by Sdyess** [comment]
So what is the goal here? Now i see changes being made for a CFBG module in the PRs? Pointless to review if you're going to take the module instead.

**Comment by Winfidonarleyan** [comment]
@Sdyess, yea

**Comment by vhiperdev** [comment]
was created a new PR I am ending this.

**Comment by BarbzYHOOL** [comment]
Thank you @vhiperdev for that CFBG and thank you Winfi for continuing!

**Comment by FrancescoBorzi** [comment]
@vhiperdev many thanks, I appreciate you sharing this regardless that it will be merged or not

**Pull request closed without merging**